### PR TITLE
Fix random Segfaults on Ubuntu arm

### LINKED ISSUE #17449: [Linux ARM] compareexchangefloat.exe intermittently fails with crossgen-d S.P.C.dll
Ubuntu Checked arm **compareexchangefloat.exe** failed 74/100 times when running using crossgen-d S.P.C.dll  and clang5.0 compiled coreclr. 

The failure could be either of those two:
```
Assert failure(PID 1923 [0x00000783], Thread: 1923 [0x0783]): sh->IsFullyInitialized()
    File: /opt/code/src/vm/safehandle.cpp Line: 65
    Image: /mnt/echesakov/coreclr/bin/tests/Linux.arm.Checked/Tests/Core_Root/corerun
```
```
Assert failure(PID 1792 [0x00000700], Thread: 1792 [0x0700]): (GetComponentSize() <= 2) || IsArray()
    File: /opt/code/src/vm/methodtable.cpp Line: 8008
    Image: /mnt/echesakov/coreclr/bin/tests/Linux.arm.Checked/Tests/Core_Root/corerun
```
It does NOT fail when S.P.C.dll is IL (at least I couldn't make it failing). 
Also not reproducible under GDB even with crossgen-d S.P.C.dll.

I suspect that we have some race condition that manifest itself only for R2R S.P.C.dll.

/cc @dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by echesakovMSFT**
Fixes #17449 #16963

/cc @sandreenko

### CODE REVIEW

File: src/pal/src/thread/context.cpp [hunk 1]
```diff
@@ -1438,8 +1438,26 @@ DBG_FlushInstructionCache(
...
+
+    do
+    {
+        pos = (begin + dwPageSize) & ~(dwPageSize - 1);
+        if (pos > end)
+            pos = end;
+
+        __builtin___clear_cache((char *)begin, (char *)pos);
+        begin = pos;
+    } while (pos < end);
```
Comment (sandreenko) [inline src/pal/src/thread/context.cpp]: why do you use loop with post condition?

**Comment by echesakovMSFT** [comment]
The following test runs with these changes and #17375 (enabling R2R `S.P.C.dll` for testing)
* [Ubuntu arm Cross Checked forcerelocs Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_forcerelocs_tst_prtest/7/consoleText)
* [Ubuntu arm Cross Checked Innerloop Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/arm_cross_checked_ubuntu_innerloop_tst_prtest/30/consoleText)
* [Ubuntu arm Cross Checked jitstress1 Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_jitstress1_tst_prtest/10/consoleText)
* [Ubuntu arm Cross Checked jitstress2 Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_jitstress2_tst_prtest/10/consoleText)
* [Ubuntu arm Cross Checked jitstressregs0x10 Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_jitstressregs0x10_tst_prtest/7/consoleText)
* [Ubuntu arm Cross Checked jitstressregs0x1000 Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_jitstressregs0x1000_tst_prtest/7/consoleText)
* [Ubuntu arm Cross Checked jitstressregs0x80 Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_jitstressregs0x80_tst_prtest/7/consoleText)
* [Ubuntu arm Cross Checked jitstressregs1 Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_jitstressregs1_tst_prtest/7/consoleText)
* [Ubuntu arm Cross Checked jitstressregs2 Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_jitstressregs2_tst_prtest/7/consoleText)
* [Ubuntu arm Cross Checked jitstressregs3 Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_jitstressregs3_tst_prtest/7/consoleText)
* [Ubuntu arm Cross Checked jitstressregs4 Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_jitstressregs4_tst_prtest/7/consoleText)
* [Ubuntu arm Cross Checked jitstressregs8 Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_jitstressregs8_tst_prtest/7/consoleText)
* [Ubuntu arm Cross Checked minopts Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_minopts_tst_prtest/10/consoleText)
* [Ubuntu arm Cross Checked normal Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/arm_cross_checked_ubuntu_tst_prtest/14/consoleText)
* [Ubuntu arm Cross Checked r2r Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/arm_cross_checked_ubuntu_r2r_tst_prtest/7/consoleText)
* [Ubuntu arm Cross Checked tailcallstress Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/jitstress/job/arm_cross_checked_ubuntu_tailcallstress_tst_prtest/7/consoleText)
* [Ubuntu arm Cross Release normal Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/arm_cross_release_ubuntu_tst_prtest/7/consoleText)
* [Ubuntu arm Cross Release r2r Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/arm_cross_release_ubuntu_r2r_tst_prtest/7/consoleText)

shows that only these two tests have failed
* CoreMangLib/system/span/SlowTailCallArgs/SlowTailCallArgs.sh in [Ubuntu arm Cross Checked r2r Build and Test](https://ci.dot.net/job/dotnet_coreclr/job/dev_unix_test_workflow/job/arm_cross_checked_ubuntu_r2r_tst_prtest/7/consoleText)
* GC/Scenarios/LeakWheel/leakwheel/leakwheel.sh failed multiple times and seems to be a known issue #16061.

No segfaults has occured during these runs though.

/cc @RussKeldorph

File: src/pal/src/thread/context.cpp (hunk 1 above)
Comment (echesakovMSFT) [reply src/pal/src/thread/context.cpp]: Did refactoring, PTAL

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Ubuntu arm Cross Checked forcerelocs Build and Test
@dotnet-bot test Ubuntu arm Cross Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstress1 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstress2 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstressregs0x10 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstressregs0x1000 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstressregs0x80 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstressregs1 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstressregs2 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstressregs3 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstressregs4 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked jitstressregs8 Build and Test
@dotnet-bot test Ubuntu arm Cross Checked minopts Build and Test
@dotnet-bot test Ubuntu arm Cross Checked normal Build and Test
@dotnet-bot test Ubuntu arm Cross Checked r2r Build and Test
@dotnet-bot test Ubuntu arm Cross Checked tailcallstress Build and Test
@dotnet-bot test Ubuntu arm Cross Release normal Build and Test
@dotnet-bot test Ubuntu arm Cross Release r2r Build and Test

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Ubuntu arm64 Cross Debug Innerloop Build

**Comment by sandreenko** [comment]
Fixes #16996, fixes #16995.

**Comment by echesakovMSFT** [comment]
@BruceForstall @jkotas @janvorli PTAL

/cc @dotnet/arm32-contrib @dotnet/jit-contrib

**Comment by echesakovMSFT** [comment]
@dotnet-bot test Ubuntu x64 Checked Innerloop Build and Test

**Comment by echesakovMSFT** [comment]
@BruceForstall Yes, I have opened another PR #17559 to track all Ubuntu/arm failing tests.

File: src/pal/src/thread/context.cpp [hunk 2]
```diff
@@ -26,6 +26,8 @@ SET_DEFAULT_DEBUG_CHANNEL(THREAD); // some headers have code with asserts, so do
 #include "pal/context.h"
 #include "pal/debug.h"
 #include "pal/thread.hpp"
+#include "pal/utils.h"
```
Comment (jaykrell) [inline src/pal/src/thread/context.cpp]: Could ifndef arm around the includes.

**Pull request merged**